platform/api/aws/iam.go

```diff
@@ -59,14 +59,50 @@ const (
 // AmazonS3RReadOnlyAccess permissions policy applied to allow fetches
 // of S3 objects that are not owned by the root account.
 func (a *API) ensureInstanceProfile(name string) error {
-	_, err := a.iam.GetInstanceProfile(&iam.GetInstanceProfileInput{
+	profileExists := false
+	policy := "AmazonS3ReadOnlyAccess"
+
+	profile, err := a.iam.GetInstanceProfile(&iam.GetInstanceProfileInput{
 		InstanceProfileName: &name,
 	})
-	if err == nil {
-		return nil
-	}
-	if awserr, ok := err.(awserr.Error); !ok || awserr.Code() != "NoSuchEntity" {
+	if awsErr, ok := err.(awserr.Error); !ok || awsErr.Code() != "NoSuchEntity" {
 		return fmt.Errorf("getting instance profile %q: %v", name, err)
+	} else if err == nil {
```

Comment:
Can't we just fold back this whole block with the rest of the logic below and just be more idempotent? E.g.:

```
- CreateInstanceProfile, ignore EntityAlreadyExists
- CreateRole, ignore EntityAlreadyExists
- PutRolePolicy (seems like that API is already idempotent)
- AddRoleToInstanceProfile, ignore EntityAlreadyExists
```

Reply:
I do like the approach, the main deviation from what you have listed is `AddRoleToInstanceProfile`, in testing this is actually not returning an `EntityAlreadyExists` but rather a `LimitExceeded` because the default limit for `InstanceSessionsPerInstanceProfile` is `1` (essentially each instance profile can only have one role).

I'm not sure I feel that error is sufficient towards ensuring that the correct `Role` is attached to the `InstanceProfile`. Perhaps I'll just add a check there in addition to ignoring the `EntityAlreadyExists` errors on `CreateInstanceProfile` & `CreateRole`